# Numpy


## Use Case: Fitness data analysis



In [1]:
!gdown 1kXqcJo4YzmwF1G2BPoA17CI49TZVHANF

Downloading...
From: https://drive.google.com/uc?id=1kXqcJo4YzmwF1G2BPoA17CI49TZVHANF
To: /content/fitness.txt
100% 3.14k/3.14k [00:00<00:00, 8.84MB/s]


In [4]:
import numpy as np

In [ ]:
data = np.loadtxt('fitness.txt', dtype='str')

In [6]:
data[:5]

array([['06-10-2017', '5464', '200', '181', '5', '0', '66'],
       ['07-10-2017', '6041', '100', '197', '8', '0', '66'],
       ['08-10-2017', '25', '100', '0', '5', '0', '66'],
       ['09-10-2017', '5461', '100', '174', '4', '0', '66'],
       ['10-10-2017', '6915', '200', '223', '5', '500', '66']],
      dtype='<U10')

What's the shape of the data?

In [ ]:
data.shape

(96, 7)

There are 96 records and each record has 7 features.
These features are:
- Date
- Step count
- Mood
- Calories Burned
- Hours of sleep
- activity status
- weight

In [ ]:
data[0]

array(['06-10-2017', '5464', '200', '181', '5', '0', '66'], dtype='<U10')

#### Whats the way to change columns to rows and rows to columns?

Transpose

In [ ]:
data.T[0]

array(['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
       '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
       '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
       '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
       '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
       '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
       '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
       '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
       '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
       '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
       '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
       '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
       '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
       '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
       '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
       '05-12-2017', '06-12-2017', '07-12-2017', '08-12

In [ ]:
(data.T)

array([['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
        '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
        '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
        '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
        '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
        '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
        '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
        '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
        '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
        '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
        '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
        '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
        '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
        '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
        '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
        '05-12-2017', '06-12-2017', '07-

In [ ]:
date, step_count, mood, calories_burned, hours_of_sleep, activity_status, weight = data.T

In [ ]:
step_count

array(['5464', '6041', '25', '5461', '6915', '4545', '4340', '1230', '61',
       '1258', '3148', '4687', '4732', '3519', '1580', '2822', '181',
       '3158', '4383', '3881', '4037', '202', '292', '330', '2209',
       '4550', '4435', '4779', '1831', '2255', '539', '5464', '6041',
       '4068', '4683', '4033', '6314', '614', '3149', '4005', '4880',
       '4136', '705', '570', '269', '4275', '5999', '4421', '6930',
       '5195', '546', '493', '995', '1163', '6676', '3608', '774', '1421',
       '4064', '2725', '5934', '1867', '3721', '2374', '2909', '1648',
       '799', '7102', '3941', '7422', '437', '1231', '1696', '4921',
       '221', '6500', '3575', '4061', '651', '753', '518', '5537', '4108',
       '5376', '3066', '177', '36', '299', '1447', '2599', '702', '133',
       '153', '500', '2127', '2203'], dtype='<U10')

In [ ]:
step_count.dtype

dtype('<U10')

#### Because Numpy type-casted all the data to strings. It's a string type where **U** means Unicode String. and 10 means 10 bytes.



**Step Count**

In [ ]:
step_count = np.array(step_count, dtype = 'int')
step_count.dtype

dtype('int32')

In [ ]:
step_count

array([5464, 6041,   25, 5461, 6915, 4545, 4340, 1230,   61, 1258, 3148,
       4687, 4732, 3519, 1580, 2822,  181, 3158, 4383, 3881, 4037,  202,
        292,  330, 2209, 4550, 4435, 4779, 1831, 2255,  539, 5464, 6041,
       4068, 4683, 4033, 6314,  614, 3149, 4005, 4880, 4136,  705,  570,
        269, 4275, 5999, 4421, 6930, 5195,  546,  493,  995, 1163, 6676,
       3608,  774, 1421, 4064, 2725, 5934, 1867, 3721, 2374, 2909, 1648,
        799, 7102, 3941, 7422,  437, 1231, 1696, 4921,  221, 6500, 3575,
       4061,  651,  753,  518, 5537, 4108, 5376, 3066,  177,   36,  299,
       1447, 2599,  702,  133,  153,  500, 2127, 2203])

**Calories Burned**

In [ ]:
calories_burned = np.array(calories_burned, dtype = 'int')
calories_burned.dtype

dtype('int32')

**Hours of Sleep**

In [ ]:
hours_of_sleep = np.array(hours_of_sleep, dtype = 'int')
hours_of_sleep.dtype

dtype('int32')

**Weight**

In [ ]:
weight = np.array(weight, dtype = 'int')
weight.dtype

dtype('int32')

**Mood**

`Mood` is a categorical data type

In [ ]:
mood

array(['200', '100', '100', '100', '200', '100', '100', '100', '100',
       '100', '100', '100', '300', '100', '100', '100', '100', '200',
       '200', '200', '200', '200', '200', '300', '200', '300', '300',
       '300', '300', '300', '300', '300', '200', '300', '300', '300',
       '300', '300', '300', '300', '300', '300', '300', '200', '300',
       '300', '300', '300', '300', '300', '300', '300', '300', '200',
       '100', '300', '300', '300', '300', '300', '300', '300', '100',
       '200', '200', '100', '100', '200', '200', '300', '200', '200',
       '100', '200', '100', '200', '200', '100', '100', '100', '100',
       '300', '200', '300', '200', '100', '100', '100', '200', '200',
       '100', '100', '300', '200', '200', '300'], dtype='<U10')

In [ ]:
np.unique(mood)

array(['100', '200', '300'], dtype='<U10')

In [ ]:
mood[mood == '300'] = 'Happy'

In [ ]:
mood[mood == '200'] = 'Neutral'

In [ ]:
mood[mood == '100'] = 'Sad'

In [ ]:
mood

array(['Neutral', 'Sad', 'Sad', 'Sad', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Sad', 'Sad', 'Sad', 'Happy', 'Sad', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral',
       'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Neutral',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Sad', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Sad', 'Neutral', 'Neutral',
       'Sad', 'Sad', 'Neutral', 'Neutral', 'Happy', 'Neutral', 'Neutral',
       'Sad', 'Neutral', 'Sad', 'Neutral', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Happy', 'Neutral', 'Happy', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Sad', 'Sad', 'Happy', 'Neutral', 'Neutral',
       'Happy'], dtype='<U10')

**Activity Status**

Here 0 means Feeling of inactiveness

500 means Feeling of activeness

In [ ]:
activity_status

array(['0', '0', '0', '0', '500', '0', '0', '0', '0', '0', '0', '0',
       '500', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0',
       '500', '0', '0', '0', '0', '500', '0', '0', '0', '0', '0', '500',
       '500', '500', '500', '500', '500', '500', '500', '500', '0', '0',
       '0', '0', '0', '0', '500', '500', '500', '500', '500', '500',
       '500', '500', '500', '500', '500', '500', '0', '500', '500', '0',
       '500', '500', '500', '500', '500', '0', '500', '500', '500', '500',
       '0', '0', '0', '0', '500', '500', '500', '500', '0', '0', '0', '0',
       '0', '0', '0', '0', '500', '0', '500'], dtype='<U10')

In [ ]:
activity_status[activity_status == '500'] = 'Active'
activity_status[activity_status == '0'] = 'Inactive'

In [ ]:
activity_status

array(['Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Active', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Inactive',
       'Inactive', 'Inactive', 'Inacti

### EDA: Insights from the data.

#### What's the average step count?

In [ ]:
step_count.mean()

2935.9375

#### On which day the step count was highest?

In [ ]:
step_count.argmax()


69

In [ ]:
date[step_count.argmax()]

'14-12-2017'

Let's check the calorie burnt on the day

In [ ]:
calories_burned[step_count.argmax()]

243

Let's try to get the number of steps on that day as well

In [ ]:
step_count.max()

7422

#### What's the most frequent mood ?

One approach is for each of the category we get count of record and see which one is the highest

In [ ]:
mood[mood == 'Sad'].shape

(29,)

In [ ]:
mood[mood == 'Neutral'].shape

(27,)

In [ ]:
mood[mood == 'Happy'].shape

(40,)

Another approach:

In [ ]:
np.unique(mood)

array(['Happy', 'Neutral', 'Sad'], dtype='<U10')

We can get the count by passing in the parameter `return_counts = True`

In [ ]:
np.unique(mood, return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'),
 array([40, 27, 29], dtype=int64))

The most frequent mood is Happy :)

#### Comparing step counts on bad mood days and good mood days

**Average step count on Sad mood days**

In [ ]:
np.mean(step_count[mood == 'Sad'])

2103.0689655172414

In [ ]:
np.sort(step_count[mood == 'Sad'])

array([  25,   36,   61,  133,  177,  181,  221,  299,  518,  651,  702,
        753,  799, 1230, 1258, 1580, 1648, 1696, 2822, 3148, 3519, 3721,
       4061, 4340, 4545, 4687, 5461, 6041, 6676])

In [ ]:
np.std(step_count[mood == 'Sad'])

2021.2355035376254

**Average step count on happy days**

In [ ]:
np.mean(step_count[mood == 'Happy'])

3392.725

In [ ]:
np.sort(step_count[mood == 'Happy'])

array([ 153,  269,  330,  493,  539,  546,  614,  705,  774,  995, 1421,
       1831, 1867, 2203, 2255, 2725, 3149, 3608, 4005, 4033, 4064, 4068,
       4136, 4275, 4421, 4435, 4550, 4683, 4732, 4779, 4880, 5195, 5376,
       5464, 5537, 5934, 5999, 6314, 6930, 7422])

Average step count on sad days - 2103.

Average step count on happy days - 3392

**There may be relation between mood and step count**

#### Let's try to check inverse. Mood when step count was greater/lesser

**Mood when step count > 4000**

In [ ]:
np.unique(mood[step_count > 4000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'),
 array([22,  9,  7], dtype=int64))

Out of 38 days when step count was more than 4000, user was feeling happy on 22 days.

**Mood when step count <= 2000**

In [ ]:
np.unique(mood[step_count < 2000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'),
 array([13,  8, 18], dtype=int64))

Out of 39 days, when step count was less than 2000, user was feeling sad on 18 days.

#### **There may be a correlation between Mood and step count**